rust_snuba/rust_arroyo/src/processing/strategies/produce.rs

```diff
@@ -0,0 +1,160 @@
+
+use log::warn;
+use crate::backends::Producer;
+use crate::backends::kafka::producer::KafkaProducer;
+use crate::backends::kafka::types::KafkaPayload;
+use crate::processing::strategies::{
+    CommitRequest, MessageRejected, ProcessingStrategy,
+};
+use crate::types::{Message, TopicOrPartition};
+use std::collections::VecDeque;
+use std::time::{Duration, Instant};
+
+pub struct Produce<TPayload: Clone + Send + Sync> {
+    pub producer: KafkaProducer,
+    pub next_step: Box<dyn ProcessingStrategy<TPayload>>,
+    queue: VecDeque<Message<TPayload>>,
+    topic: TopicOrPartition,
+    closed: bool,
+    max_queue_size: usize,
+}
+
+// TODO: make the queue store futures
+// TODO: make this generic
+impl Produce<KafkaPayload> {
+    pub fn new(producer: KafkaProducer, next_step: Box<dyn ProcessingStrategy<KafkaPayload>>, topic:TopicOrPartition) -> Self {
+        Produce {
+            producer,
+            next_step,
+            queue: VecDeque::new(),
+            topic,
+            closed: false,
+            max_queue_size: 1000,
+        }
+    }
+}
+
+impl ProcessingStrategy<KafkaPayload>
+    for Produce<KafkaPayload>
+{
+    fn poll(&mut self) -> Option<CommitRequest> {
+        while !self.queue.is_empty(){
+            let message = self.queue.pop_front();
+            self.next_step.poll();
+            self.next_step.submit(message.unwrap()).unwrap()
+        }
+        return None
+    }
+
+    fn submit(&mut self, message: Message<KafkaPayload>) -> Result<(), MessageRejected> {
+        if self.queue.len() >= self.max_queue_size {
+            return Err(MessageRejected)
+        }
+        print!("Producing message: {:?}", &message.payload.payload);
+        self.producer.produce(&self.topic, &message.payload);
+        self.producer.flush();
```

Comment:
`submit` is supposed to be a non blocking method, here we're waiting indefinitely for the message to get flushed

Reply:
removed flush and made it place a future on the queue.